### DISCUSSION TITLE
pooling engine instances for better user  query experience

### DISCUSSION THREAD

**Discussion by zinking** [apache/kyuubi#1929] [Ideas]
the amount of time it takes to bootstrap a spark engine instance is currently innegligible.  30s at minimum. 
it is possible to pool a few spark engine instances at the server and then later assigning to user when requests come in.

this will cause a certain a mount of resource waste, but with DRA and controlling the pool size , waste could be reduced.

**Comment by yaooqinn** [comment]
you can create a dummy connection for that user before he/she do the real work

**Reply by zinking** [reply]
don't really know who he/she is, I mean this could be a built-in mechanism for bringing engine up.  would this be helpful ?

**Reply by yaooqinn** [reply]
the engine is user-oriented. mostly you need to start an engine for a specific user.

if you are not working with a secured cluster and want all engines to be shared across users, you can use GROUP or SERVER with engine-pool together.

**Reply by zinking** [reply]
sounds good. but is group/server level engine instance pooled already ? I thought they had only 1 instance as well and not pooled.

**Reply by yaooqinn** [reply]
yes, the engine instance is pooled(mostly), and `kyuubi.engine.pool.size` can help create multi instances inside that pool

**Reply by zinking** [reply]
thanks for the explaination

**Comment by pan3793** [comment]
Kyuubi provides a conf `kyuubi.session.engine.launch.async`, default is `false`, does it help?

**Reply by zinking** [reply]
I investigated a little bit on this option. I can't figure out the essence of async here. is it `async` only meaningful to the rest client ?  I mean when we open a session from beeline.  `async` or **not** seems not a concern as we all have to wait till the session is up. 

what I mentioned above is we have some sessions brought up when server is started so users don't have to wait for the sparksession to start up. as per my understanding open session is triggered when user try to connects to the server for the first time.

**Reply by pan3793** [reply]
> is it async only meaningful to the rest client ?

No, it's for thrift client, jdbc/beeline can benefit it.

The async mode can not shorten the spark driver launch time, the benefits are:
1) when client connect to kyuubi server, it return immediately no matter engine is ready or not, and pick or launch the engine in the background, so in client view, the open session step always fast, but the sql will be blocked until the engine launched.
2) the server can echo logs of launch engine to the client, e.g. `spark-submit` commands

**Reply by zinking** [reply]
I guess this should be the default implementation then.  

and from the user's perspective, it is not **async**  because now the engine start up times gets diluted in the calculation step.

guess I mis-understood @yaooqinn 's explanation as well. because I'm targeting shortening spark launch time here.

**Reply by pan3793** [reply]
You are right. `async` has been introduced in v1.4.0 and default enabled since v1.5.0

**Reply by pan3793** [reply]
> I'm targeting shortening spark launch time here.

Are you using Spark on Yarn?

**Reply by zinking** [reply]
yes, I was suggesting pooling say 1 spark instance when kyuubi server starts, and when user comes in we allocate that instance to the user and refill that pool. 

but this seems rather complicated in the current ha setup.

**Reply by pan3793** [reply]
You can pre-install Spark binaries on Yarn workers, and configure `spark.yarn.jars`, it can save time of uploading and distributing Spark jars of launching engine
https://spark.apache.org/docs/latest/running-on-yarn.html
